lisa/tests/staging/scheduler/util_tracking.py

```diff
@@ -0,0 +1,377 @@
+        pl.savefig(figname, bbox_inches='tight')
+        plt.close()
+
+    @requires_events('sched_util_est_task', 'sched_load_se')
+    @RTAEventsAnalysis.task_phase_windows.used_events
+    def test_areas(self) -> ResultBundle:
+        """
+        Test signals are properly "dominated".
+
+        The integral of `util_est_enqueued` is expected to be always not
+        smaller than that of `util_avg`, since this last is subject to decays
+        while the first not.
+
+        On fast-ramp systems, the `util_est_ewma` signal is never smaller then
+        the `util_est_enqueued`, thus his integral is expected to be bigger.
+
+        On non fast-ramp systems instead, the `util_est_ewma` is expected to be
+        smaller then `util_est_enqueued` in ramp-up phases, or bigger in
+        ramp-down phases.
+
+        Those conditions are checked on a single execution of a task which has
+        three main behaviours:
+        - STABLE: periodic big task running for a relatively long period to
+                  ensure `util_avg` saturation.
+        - DOWN: periodic ramp-down task, to slowly decay `util_avg`
+        - UP: periodic ramp-up task, to slowly increase `util_avg`
+
+        """
+        PhaseStats = namedtuple("PhaseStats", [
+            'start', 'end', 'area_util', 'area_enqueued', 'area_ewma'])
+        failure_reasons = {}
+        metrics = {}
+
+        # We have only two task: the main 'rt-app' task and our 'test_task'
+        test_task = self.trace.analysis.rta.rtapp_tasks[-1]
+
+        ue_df = self.trace.df_events('sched_util_est_task')
+        ue_df = ue_df[ue_df.pid == test_task.pid]
+        ua_df = self.trace.df_events('sched_load_se')
+        ua_df = ua_df[ua_df.pid == test_task.pid]
+
+        passed = True
+        for phase in self.trace.analysis.rta.task_phase_windows(test_task):
+            phase_df = ue_df[phase.start:phase.end]
+            area_enqueued = series_integrate(phase_df.util_est_enqueued)
+            area_ewma = series_integrate(phase_df.util_est_ewma)
+
+            phase_df = ua_df[phase.start:phase.end]
+            area_util = series_integrate(phase_df.util)
+
+            metrics[phase.id] = PhaseStats(phase.start, phase.end,
+                                           area_util, area_enqueued, area_ewma)
+
+            if area_enqueued < area_util:
+                failure_reasons[phase.id] = 'Enqueued smaller then Util Average'
+                passed = False
+
+            # Running on FastRamp kernels:
+            if self.fast_ramp:
+
+                # STABLE, DOWN and UP:
+                if area_ewma < area_enqueued:
+                    failure_reasons[phase.id] = 'NO_FAST_RAMP: EWMA smaller then Enqueued'
+                    passed = False
+
+            # Running on (legacy) non FastRamp kernels:
+            else:
+
+                # STABLE: ewma ramping up
+                if phase.id == 0 and area_ewma > area_enqueued:
```

Comment:
@valschneider would be great to be able to name the phases in the profile, so rta analysis can resolve the original name for each phase given an rtapp profile instead of just giving back a numeric ID

Reply:
@derkling is the id returned by the rta analysis stable when the rtapp phase is looping multiple times, or will it keep increasing monotonically ?

Reply:
Hmm one thing we could do is perhaps add some mapping of `name : id`, or more globally give the same name to all phases of a given `RTATask` but create disjoined mappings when combining `RTATasks`.

For instance you could have `task = Periodic(..., "setup") + Periodic(..., "work")` and you'd get something like `task.phase_regions` which would be ~ `{"setup" : [<phases of the first Periodic incantation>], "work" : [<phases of the second periodic incantation>]}`

Reply:
Good point, that's rt-app specific... do you mean when a single phase is iterated multiple times?
Not even sure we can have that in RTApp... I mean, you can have "phase's loops", but I don't think you can have "loop of the same phases", that would be essentially the same as the first.

Reply:
Ok, so as long as the loop iterations in rtapp as considered as being part of a given phase ID, we can compute quite easily the mapping from names to phase IDs. The mapping would need to allocate IDs sequentially though, since it has to match what the new rta analysis gives (that's the whole point).